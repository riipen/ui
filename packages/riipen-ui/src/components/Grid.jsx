import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";

const SPACINGS = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10];

class Grid extends React.Component {
  static propTypes = {
    /**
     * A whitelisted set of align items options for the grid.
     */
    alignItems: PropTypes.oneOf([
      "flex-start",
      "center",
      "flex-end",
      "stretch",
      "baseline"
    ]),

    /**
     * The content of the grid.
     */
    children: PropTypes.any,

    /**
     * An array of custom CSS classes to apply.
     */
    classes: PropTypes.array,

    /**
     * The styling to pass into the flex-direction of the grid.
     */
    flexDirection: PropTypes.oneOf([
      "row",
      "row-reverse",
      "column",
      "column-reverse",
      "initial",
      "inherit"
    ]),

    /**
     * Defines if the grid should grow to fill the space available
     */
    flexGrow: PropTypes.number,

    /**
     * Defines if the grid should shrink to allow more space available
     */
    flexShrink: PropTypes.number,

    /**
     * The styling to pass into the flex-wrap of the grid.
     */
    flexWrap: PropTypes.oneOf([
      "nowrap",
      "wrap",
      "wrap-reverse",
      "inherit",
      "initial",
      "unset"
    ]),

    /**
     * A whitelisted set of justify content options for the grid.
     */
    justifyContent: PropTypes.oneOf([
      "flex-start",
      "center",
      "flex-end",
      "space-between",
      "space-around",
      "space-evenly"
    ]),

    /**
     * Defines the space between grid items in the grid.
     */
    spacing: PropTypes.oneOf(SPACINGS),

    /**
     * The function to callback to when the grid changes sizes
     * The callback includes the current size of the grid.
     */
    onResize: PropTypes.func
  };

  static defaultProps = {
    alignItems: "flex-start",
    classes: [],
    justifyContent: "flex-start",
    spacing: 3,
    flexGrow: 1,
    flexShrink: 1,
    flexDirection: "row",
    flexWrap: "wrap",
    onResize: () => {}
  };

  constructor(props) {
    super(props);
    this.state = {
      size: "lg"
    };
    this.resizeRef = React.createRef();
  }

  componentDidMount = () => {
    // If the user has support for the ResizeObserver use that, if not use the window resize event

    if (typeof ResizeObserver !== "undefined") {
      this.resizeObserver = new ResizeObserver(this.onResize);

      if (this.resizeObserver && this.resizeRef.current) {
        this.resizeObserver.observe(this.resizeRef.current);
      }
    } else {
      window.addEventListener("resize", this.onResize);
    }

    this.onResize();
  };

  componentWillUnmount = () => {
    if (this.resizeObserver) {
      this.resizeObserver.disconnect();
    } else {
      window.removeEventListener("resize", this.onResize);
    }
  };

  onResize = () => {
    const size = this.getComponentSize();
    this.setState({ size });
    this.props.onResize(size);
  };

  getComponentSize = () => {
    const width = this.resizeRef.current?.clientWidth || 1080;
    return (
      (width >= this.context.breakpoints.lg && "lg") ||
      (width >= this.context.breakpoints.md && "md") ||
      (width >= this.context.breakpoints.sm && "sm") ||
      "xs"
    );
  };

  static contextType = ThemeContext;

  render() {
    const {
      alignItems,
      children,
      classes,
      justifyContent,
      spacing,
      flexDirection,
      flexGrow,
      flexShrink,
      flexWrap
    } = this.props;

    const { size } = this.state;

    const theme = this.context;

    const className = clsx(classes, size);

    const childrenWithProps = React.Children.map(children, child => {
      if (!child) return null;

      return React.cloneElement(child, { spacing, size });
    });

    return (
      <React.Fragment>
        <div ref={this.resizeRef} className={className}>
          {childrenWithProps}
        </div>
        <style jsx>{`
          div {
            align-items: ${alignItems};
            display: flex;
            flex-direction: ${flexDirection};
            flex-grow: ${flexGrow};
            flex-shrink: ${flexShrink};
            flex-wrap: ${flexWrap};
            justify-content: ${justifyContent};
            margin-bottom: -${theme.spacing(spacing)}px;
            margin-left: -${theme.spacing(spacing)}px;
            min-height: 0;
            position: relative;
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default withClasses(Grid);
