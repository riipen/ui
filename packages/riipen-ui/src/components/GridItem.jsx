import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";

const COLUMNS = 12;
const SPACINGS = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10];

class GridItem extends React.Component {
  static propTypes = {
    /**
     * The content of the grid item.
     */
    children: PropTypes.any,

    /**
     * An array of custom CSS classes to apply.
     */
    classes: PropTypes.array,

    /**
     * Defines if the grid item should grow to fill extra space in the row
     */
    flexGrow: PropTypes.bool,

    /**
     * The columns use at the large resolution. Can also be 'hidden'.
     */
    lg: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),

    /**
     * The columns use at the medium resolution. Can also be 'hidden'.
     */
    md: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),

    /**
     * The columns use at the small resolution. Can also be 'hidden'.
     */
    sm: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),

    /**
     * Defines the space between this grid item and other items.
     */
    spacing: PropTypes.oneOf(SPACINGS)
  };

  static defaultProps = {
    classes: [],
    lg: 12,
    spacing: 3,
    flexGrow: false
  };

  static contextType = ThemeContext;

  render() {
    const { children, classes, lg, md, sm, spacing, flexGrow } = this.props;

    const theme = this.context;

    const className = clsx(classes);

    const grow = flexGrow ? 1 : 0;

    return (
      <React.Fragment>
        <div className={className}>{children}</div>
        <style jsx>{`
          div {
            box-sizing: border-box;
            margin-bottom: ${theme.spacing(spacing)}px;
            padding-left: ${theme.spacing(spacing)}px;
            flex: ${grow} 0 ${(+lg / COLUMNS) * 100 || 0}%;
          }
          @media (max-width: ${theme.breakpoints.md}px) {
            div {
              display: ${md === "hidden" ? "none" : "initial"};
              flex: ${grow} 0 ${((+md || +lg) / COLUMNS) * 100 || 0}%;
            }
          }
          @media (max-width: ${theme.breakpoints.sm}px) {
            div {
              display: ${[sm, md].includes("hidden") ? "none" : "initial"};
              flex: ${grow} 0 ${((+sm || +md || +lg) / COLUMNS) * 100 || 0}%;
            }
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default GridItem;
