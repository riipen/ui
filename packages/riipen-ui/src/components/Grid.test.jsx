import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import defaultTheme from "../themes/default";

import Grid from "./Grid.jsx";
import GridItem from "./GridItem.jsx";

describe("<Grid>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(<Grid />);
    } catch (e) {
      error = e;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(<Grid />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("default props", () => {
    it("sets correct default props", () => {
      const defaultProps = new Grid().type.defaultProps;

      const wrapper = mount(<Grid />);

      const component = wrapper.find("Grid");
      expect(component.props().alignItems).toEqual(defaultProps.alignItems);
      expect(component.props().justifyContent).toEqual(
        defaultProps.justifyContent
      );
      expect(component.props().spacing).toEqual(defaultProps.spacing);
      expect(component.props().flexGrow).toEqual(defaultProps.flexGrow);
      expect(component.props().flexShrink).toEqual(defaultProps.flexShrink);
      expect(component.props().flexDirection).toEqual(
        defaultProps.flexDirection
      );
      expect(component.props().flexWrap).toEqual(defaultProps.flexWrap);
      expect(component.props().onResize).toEqual(defaultProps.onResize);
    });
  });

  describe("alignItems prop", () => {
    it("sets given valid custom alignItems", () => {
      const alignItems = "stretch";

      const wrapper = mount(<Grid alignItems={alignItems} />);

      expect(wrapper.find("JSXStyle").props().dynamic).toContain(alignItems);
    });

    it("gives an error given an invalid alignItems", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();
      const alignItems = "middle";

      mount(<Grid alignItems={alignItems} />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });

  describe("children prop", () => {
    it("renders given non null children", () => {
      const child = <GridItem />;

      const wrapper = mount(
        <Grid>
          {null}
          {child}
        </Grid>
      );

      expect(
        wrapper
          .find("Grid")
          .childAt(0)
          .containsMatchingElement(child)
      ).toEqual(true);
    });

    it("passes spacing to children nodes", () => {
      const child = <GridItem />;
      const spacing = 0;

      const wrapper = mount(<Grid spacing={spacing}>{child}</Grid>);

      expect(wrapper.find("GridItem").props().spacing).toEqual(spacing);
    });

    it("passes size to children nodes", () => {
      const size = "sm";
      const child = <GridItem />;

      const wrapper = mount(<Grid>{child}</Grid>);

      wrapper.find("Grid").setState({ size });

      expect(wrapper.find("GridItem").props().size).toEqual(size);
    });
  });

  describe("classes prop", () => {
    it("applies custom classes", () => {
      const classes = ["classOne"];

      const wrapper = mount(<Grid classes={classes} />);

      expect(
        wrapper
          .find("Grid")
          .childAt(0)
          .hasClass(classes[0])
      ).toEqual(true);
    });

    it("appends higher order values to classes prop with withClass decorator", () => {
      const classes = ["riipen", "riipen-grid"];

      const wrapper = mount(<Grid />);

      expect(
        wrapper
          .find("Grid")
          .props()
          .classes.sort()
      ).toEqual(classes.sort());
    });
  });

  describe("flexDirection prop", () => {
    it("sets given valid custom flexDirection", () => {
      const flexDirection = "column-reverse";

      const wrapper = mount(<Grid flexDirection={flexDirection} />);

      expect(wrapper.find("JSXStyle").props().dynamic).toContain(flexDirection);
    });

    it("gives an error given an invalid flexDirection", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();
      const flexDirection = "middle";

      mount(<Grid flexDirection={flexDirection} />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });

  describe("flexGrow prop", () => {
    it("sets given custom flexGrow", () => {
      const flexGrow = 100;

      const wrapper = mount(<Grid flexGrow={flexGrow} />);

      expect(wrapper.find("JSXStyle").props().dynamic).toContain(flexGrow);
    });
  });

  describe("flexShrink prop", () => {
    it("sets given custom flexShrink", () => {
      const flexShrink = 100;

      const wrapper = mount(<Grid flexShrink={flexShrink} />);

      expect(wrapper.find("JSXStyle").props().dynamic).toContain(flexShrink);
    });
  });

  describe("flexWrap prop", () => {
    it("sets given custom flexWrap", () => {
      const flexWrap = "unset";

      const wrapper = mount(<Grid flexWrap={flexWrap} />);

      expect(wrapper.find("JSXStyle").props().dynamic).toContain(flexWrap);
    });

    it("gives an error given an invalid flexWrap", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();
      const flexWrap = "middle";

      mount(<Grid flexWrap={flexWrap} />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });

  describe("justifyContent prop", () => {
    it("sets given custom justifyContent", () => {
      const justifyContent = "center";

      const wrapper = mount(<Grid justifyContent={justifyContent} />);

      expect(wrapper.find("JSXStyle").props().dynamic).toContain(
        justifyContent
      );
    });

    it("gives an error given an invalid justifyContent", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();
      const justifyContent = "middle";

      mount(<Grid justifyContent={justifyContent} />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });

  describe("spacing prop", () => {
    it("sets given custom spacing", () => {
      const spacing = 10;

      const wrapper = mount(<Grid spacing={spacing} />);

      expect(wrapper.find("JSXStyle").props().dynamic).toContain(
        defaultTheme.spacing(spacing)
      );
    });

    it("gives an error given an invalid spacing", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();
      const spacing = 20;

      mount(<Grid spacing={spacing} />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });

  describe("onResize prop", () => {
    it("invokes onResize function on mount", () => {
      const onResize = jest.fn();

      mount(<Grid onResize={onResize} />);

      expect(onResize).toHaveBeenCalledTimes(1);
    });
  });
});
