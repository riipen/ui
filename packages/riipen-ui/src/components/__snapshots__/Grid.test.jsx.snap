// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Grid> renders correct snapshot 1`] = `
<withClasses(Grid)>
  <Grid
    alignItems="flex-start"
    classes={
      Array [
        "riipen",
        "riipen-grid",
      ]
    }
    flexDirection="row"
    flexGrow={1}
    flexShrink={1}
    flexWrap="wrap"
    justifyContent="flex-start"
    onResize={[Function]}
    spacing={3}
  >
    <div
      className="jsx-738249180 riipen riipen-grid md"
    />
    <JSXStyle
      dynamic={
        Array [
          "flex-start",
          "row",
          1,
          1,
          "wrap",
          "flex-start",
          15,
          15,
        ]
      }
      id="3466573101"
    />
  </Grid>
</withClasses(Grid)>
`;
